README
IPC::PrettyPipe

NAME

    IPC::PrettyPipe - human readable command pipelines

DESCRIPTION

    IPC::PrettyPipe provides a mechanism to maintain human readable
    execution pipelines.  Pipelines are created by adding commands to
    a IPC::PrettyPipe object.  Options to the commands are set using a
    readable format; IPC::PrettyPipe takes care of quoting strings,
    sticking equal signs in, etc.  The pipeline may be rendered into a
    nicely formatted string, as well as being executed.


DEPENDENCIES

    None

INSTALLATION


    To compile, test and install:

	perl Makefile.PL
	make
	make test
	make install


CHANGES

    See the Changes file for more details

COPYRIGHT

--8<--8<--8<--8<--

Copyright (C) 2014 Smithsonian Astrophysical Observatory

This file is part of IPC::PrettyPipe

IPC::PrettyPipe is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

-->8-->8-->8-->8--


AUTHOR

    Diab Jerius (djerius@cpan.org)



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.13      2020-07-08 17:28:45-04:00 America/New_York

    [API BREAKAGE]

      * Constructors for render and execute classes are no longer
        passed an IPC::PrettyPipe object.  It was poor design to do
        so, and was because I was the wrong kind of lazy.  Instead,
        the pipe object is passed to the render/run methods.

    [ENHANCEMENTS]

      * |= and | operators are overloaded for pipes and commands
        to make it easier to construct pipes.

      * nested pipes are by default merged if possible.
        the IPC::PrettyPipe->merge_pipes attribute controls this.

      * new render module, Struct, renders the pipeline as Perl
        structures.  Useful for testing pipeline generation.

      * The Text::Template renderer constructor can be passed
        the colorize attribute.

0.12      2018-12-21 14:02:48-05:00 America/New_York

      * No changes from previous release.
      * Promoted to stable.


0.11      2018-12-20 17:10:44-05:00 America/New_York (TRIAL RELEASE)

  [BUILD FIX]

     * previous release had an incomplete Makefile.PL, causing
       failures on Windows

0.10      2018-12-19 12:43:10-05:00 America/New_York (TRIAL RELEASE)


   [TEST FIX]

      * fix syntax in tests for Perl <= 5.16.3  (RT#128069, ++srezic)

0.09      2018-12-19 12:43:10-05:00 America/New_York (TRIAL RELEASE)


   [ENHANCEMANT & API BREAKAGE]

      * pipes may now be nested. Not all nested pipes may be
        executed, though; see the docs.
      * IPC::PrettyPipe::Render::Template::Tiny now requires two
        templates to handle nested pipes.  This breaks its API


0.08      2018-11-16 14:27:40-05:00 America/New_York


      * No changes from previous release.
      * Promote to stable.


0.07      2018-11-15 10:03:28-05:00 America/New_York (TRIAL RELEASE)


  [BUILD FIX]

     * previous release had an incomplete Makefile.PL, causing
       failures on Windows

     * note to self: don't release under Dist::Zilla::Shell.
       Dist::Zilla::Plugin::DynamicPrereqs isn't designed to be run in
       that environment.  It munges Makefile.PL and caches what it has
       done to avoid duplication. Unfortunately, this means that
       Makefile.PL will not be properly munged after the first build.


0.06      2018-11-09 16:24:42-05:00 America/New_York (TRIAL RELEASE)

  [BUILD FIX]

     * test dependencies were incompletely listed in dist.ini
     * tests now moved as much as possible to Test2::V0


0.05      2018-11-09 16:24:42-05:00 America/New_York (TRIAL RELEASE)

  [BUG FIX ]

     * Fix RT# 108397. Finally. (srezic++)


0.04      2018-11-07 17:01:33-05:00 America/New_York (TRIAL RELEASE)

  [BUILD FIX]

     * switch from Module::Install to Dist::Zilla + EU::MM


0.03_03 2014-08-05T12:58:51-0400
  [BUG FIX]

     * make work on Perl 5.10.0; see
       http://www.cpantesters.org/cpan/report/15ef7e40-150f-11e4-b698-db7a2867dcfa

0.03_02 2014-07-21T17:29:59-0400

  [BUILD FIX]

     * remove extraneous dependency on Exporter::Tiny. Jumped the gun
       on
       http://www.cpantesters.org/cpan/report/0c9e658d-6c00-1014-9524-043f55012b15,
       which reported a problem, but it's with Type::Tiny; this module
       doesn't doesn't explicitly use it.  Might be a tester problem.


0.03_01 2014-07-21T17:21:35-0400

  [BUILD FIX]
     * add missing dependency on Exporter::Tiny

0.03	2014-07-18T16:42:33-0400


  [BUG FIX]

    - renderers/executors provided as class names were not properly
      handled

    - adding a stream to a pipe object was broken

    - the Template::Tiny renderer didn't properly quote arguments for
      the shell

  [ENHANCEMENT]

    - commands, attribute names, and attribute values now have a
      "quoted_" accessor which returns strings properly escaped for
      use by a shell.

  [BUILD FIX]

    - add missing runtime dependency on List::MoreUtils


0.02	2014-02-14T15:49:54-0500

    - promote to stable, CPAN Testers is all green!

0.01_03	2014-02-09T07:53:24-0500

  [BUILD FIX]
    - require IO::ReStoreFH >= 0.04 to handle Perl 5.10.x/FileHandle problems
    - require Win32::Console::ANSI on Win32 boxes

0.01_02	 2014-02-07T06:15:35-0500

  [TEST FIX]
    - IPC::Run can't redirect fd's > 2 on Win32; skip tests which do
      that

0.01_01	2014-02-05T18:59:08-0500

  [BUG FIX]
    - remove errant 'use DDP' statement

  [OTHER]
    - switch from Probe::Perl to Devel::FindPerl


0.01  	2014-02-04T17:54:09-0500

    - first public release


0.00	Mon Sep 10 11:36:14 EDT 2012

    - forked from private IPC::PipeC module


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.014.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "manage human readable external command execution pipelines",
  "AUTHOR" => "Diab Jerius <djerius\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "IPC-PrettyPipe",
  "LICENSE" => "gpl",
  "NAME" => "IPC::PrettyPipe",
  "PREREQ_PM" => {
    "IO::ReStoreFH" => "0.04",
    "IPC::Run" => "20180523.0",
    "List::MoreUtils" => 0,
    "List::Util" => "1.34",
    "Module::Load" => 0,
    "Module::Runtime" => 0,
    "Moo" => "1.001000",
    "MooX::Attributes::Shadow" => "0.05",
    "Safe::Isa" => 0,
    "String::ShellQuote" => 0,
    "Template::Tiny" => 0,
    "Term::ANSIColor" => 0,
    "Try::Tiny" => 0,
    "Type::Tiny" => "0.038",
    "parent" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "Devel::FindPerl" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Slurper" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "File::pushd" => 0,
    "Getopt::Long" => 0,
    "IO::File" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Moo" => "1.001000",
    "Scalar::Util" => 0,
    "Template::Tiny" => 0,
    "Term::ANSIColor" => 0,
    "Test2::API" => 0,
    "Test2::V0" => 0,
    "Test::Lib" => 0,
    "Test::More" => 0,
    "Test::Trap" => 0,
    "Types::Standard" => 0
  },
  "VERSION" => "0.14",
  "test" => {
    "TESTS" => "t/*.t t/arg/*.t t/cmd/*.t t/dsl/*.t t/overload/*.t t/ppipe/*.t t/render/*.t t/run/*.t t/stream/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Dumper" => 0,
  "Devel::FindPerl" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Slurper" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "File::pushd" => 0,
  "Getopt::Long" => 0,
  "IO::File" => 0,
  "IO::Handle" => 0,
  "IO::ReStoreFH" => "0.04",
  "IPC::Open3" => 0,
  "IPC::Run" => "20180523.0",
  "List::MoreUtils" => 0,
  "List::Util" => "1.34",
  "Module::Load" => 0,
  "Module::Runtime" => 0,
  "Moo" => "1.001000",
  "MooX::Attributes::Shadow" => "0.05",
  "Safe::Isa" => 0,
  "Scalar::Util" => 0,
  "String::ShellQuote" => 0,
  "Template::Tiny" => 0,
  "Term::ANSIColor" => 0,
  "Test2::API" => 0,
  "Test2::V0" => 0,
  "Test::Lib" => 0,
  "Test::More" => 0,
  "Test::Trap" => 0,
  "Try::Tiny" => 0,
  "Type::Tiny" => "0.038",
  "Types::Standard" => 0,
  "parent" => 0
);

# inserted by Dist::Zilla::Plugin::DynamicPrereqs 0.036
if ($^O =~ /Win32/i) {
requires( 'Win32::Console::ANSI' )
}



unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

# inserted by Dist::Zilla::Plugin::DynamicPrereqs 0.036
sub _add_prereq {
  my ($mm_key, $module, $version_or_range) = @_;
  $version_or_range ||= 0;
  warn "$module already exists in $mm_key (at version $WriteMakefileArgs{$mm_key}{$module}) -- need to do a sane metamerge!"
    if exists $WriteMakefileArgs{$mm_key}{$module}
      and $WriteMakefileArgs{$mm_key}{$module} ne '0'
      and $WriteMakefileArgs{$mm_key}{$module} ne $version_or_range;
  warn "$module already exists in FallbackPrereqs (at version $FallbackPrereqs{$mm_key}{$module}) -- need to do a sane metamerge!"
    if exists $FallbackPrereqs{$module} and $FallbackPrereqs{$module} ne '0'
        and $FallbackPrereqs{$module} ne $version_or_range;
  $WriteMakefileArgs{$mm_key}{$module} = $FallbackPrereqs{$module} = $version_or_range;
  return;
}

sub requires { goto &runtime_requires }

sub runtime_requires {
  my ($module, $version_or_range) = @_;
  _add_prereq(PREREQ_PM => $module, $version_or_range);
}
